File: shared/mqtt/topics.py
"""兼容层：基于 ``TopicManager`` 的旧 Topic 工具 API。"""

from typing import Optional

from shared.mqtt.topic_manager import TopicManager, TopicType


_TOPIC_MANAGER = TopicManager()


class TopicBuilder:
    """Topic 构建器兼容封装。"""

    @staticmethod
    def control(room_id: str, agent_id: str) -> str:
        return _TOPIC_MANAGER.build_control_topic(room_id, agent_id)

    @staticmethod
    def state(room_id: str, agent_id: str) -> str:
        return _TOPIC_MANAGER.build_state_topic(room_id, agent_id)

    @staticmethod
    def describe(room_id: str, agent_id: str) -> str:
        return _TOPIC_MANAGER.build_describe_topic(room_id, agent_id)

    @staticmethod
    def description(room_id: str, agent_id: str) -> str:
        return _TOPIC_MANAGER.build_description_topic(room_id, agent_id)

    @staticmethod
    def heartbeat(room_id: str, agent_id: str) -> str:
        return _TOPIC_MANAGER.build_heartbeat_topic(room_id, agent_id)

    @staticmethod
    def global_state() -> str:
        return _TOPIC_MANAGER.build_global_state_topic()

    @staticmethod
    def policy() -> str:
        return _TOPIC_MANAGER.build_policy_topic()

    @staticmethod
    def arbitration(request_id: Optional[str] = None) -> str:
        if request_id:
            return _TOPIC_MANAGER.build_arbitration_response_topic(request_id)
        return _TOPIC_MANAGER.build_arbitration_topic()

    @staticmethod
    def events() -> str:
        return _TOPIC_MANAGER.build_events_topic()

    @staticmethod
    def system_discovery(room_id: str) -> str:
        return f"room/{room_id}/system/discovery"

    @staticmethod
    def system_error(room_id: str) -> str:
        return f"room/{room_id}/system/error"


class TopicParser:
    """Topic 解析器兼容封装。"""

    @staticmethod
    def parse(topic: str) -> dict:
        info = _TOPIC_MANAGER.parse_topic(topic)
        if info is None:
            return {}

        message_type_map = {
            TopicType.GLOBAL_STATE: "state",
            TopicType.ARBITRATION_RESPONSE: "arbitration",
        }

        result = {
            "type": info.scope,
            "room_id": info.room_id,
            "agent_id": info.agent_id,
            "message_type": message_type_map.get(info.topic_type, info.topic_type.value),
        }

        if info.topic_type == TopicType.ARBITRATION_RESPONSE:
            result["is_response"] = True
            result["request_id"] = info.correlation_id

        return result


class QoSConfig:
    """QoS 配置兼容封装。"""

    QOS_MAP = {
        "control": TopicType.CONTROL,
        "state": TopicType.STATE,
        "describe": TopicType.DESCRIBE,
        "description": TopicType.DESCRIPTION,
        "heartbeat": TopicType.HEARTBEAT,
        "home/state": TopicType.GLOBAL_STATE,
        "home/policy": TopicType.POLICY,
        "home/arbitration": TopicType.ARBITRATION,
        "home/arbitration/response": TopicType.ARBITRATION_RESPONSE,
        "home/events": TopicType.EVENTS,
    }

    @classmethod
    def get_qos(cls, message_type: str) -> int:
        topic_type = cls.QOS_MAP.get(message_type)
        if topic_type is None:
            return 0
        return _TOPIC_MANAGER.get_qos_for_topic(topic_type)

    @classmethod
    def get_qos_for_topic(cls, topic: str) -> int:
        info = _TOPIC_MANAGER.parse_topic(topic)
        if info is None:
            return 0
        return _TOPIC_MANAGER.get_qos_for_topic(info.topic_type)


class SubscriptionTopics:
    """订阅 Topic 模式兼容封装。"""

    @staticmethod
    def personal_agent(room_id: str) -> list[str]:
        return [
            _TOPIC_MANAGER.build_wildcard_topic(room_id, TopicType.STATE),
            _TOPIC_MANAGER.build_wildcard_topic(room_id, TopicType.DESCRIPTION),
            _TOPIC_MANAGER.build_global_state_topic(),
            _TOPIC_MANAGER.build_policy_topic(),
            "home/arbitration/response/+",
        ]

    @staticmethod
    def room_agent(room_id: str, agent_id: str) -> list[str]:
        return [
            _TOPIC_MANAGER.build_control_topic(room_id, agent_id),
            _TOPIC_MANAGER.build_describe_topic(room_id, agent_id),
            _TOPIC_MANAGER.build_policy_topic(),
        ]

    @staticmethod
    def central_agent() -> list[str]:
        return [
            "room/+/agent/+/state",
            "room/+/agent/+/heartbeat",
            _TOPIC_MANAGER.build_arbitration_topic(),
        ]

File: shared/js/topics.js
/**
 * A2A Topic 命名空间工具类
 * 
 * 根据 docs/communication.md 规范定义 Topic 命名空间
 */

export class TopicBuilder {
  /**
   * 控制命令 Topic
   * @param {string} roomId - 房间 ID
   * @param {string} agentId - Agent ID
   * @returns {string} Topic: room/{room_id}/agent/{agent_id}/control
   */
  static control(roomId, agentId) {
    return `room/${roomId}/agent/${agentId}/control`
  }

  /**
   * 状态发布 Topic
   * @param {string} roomId - 房间 ID
   * @param {string} agentId - Agent ID
   * @returns {string} Topic: room/{room_id}/agent/{agent_id}/state
   */
  static state(roomId, agentId) {
    return `room/${roomId}/agent/${agentId}/state`
  }

  /**
   * 能力查询 Topic
   * @param {string} roomId - 房间 ID
   * @param {string} agentId - Agent ID
   * @returns {string} Topic: room/{room_id}/agent/{agent_id}/describe
   */
  static describe(roomId, agentId) {
    return `room/${roomId}/agent/${agentId}/describe`
  }

  /**
   * 能力响应 Topic
   * @param {string} roomId - 房间 ID
   * @param {string} agentId - Agent ID
   * @returns {string} Topic: room/{room_id}/agent/{agent_id}/description
   */
  static description(roomId, agentId) {
    return `room/${roomId}/agent/${agentId}/description`
  }

  /**
   * 心跳 Topic
   * @param {string} roomId - 房间 ID
   * @param {string} agentId - Agent ID
   * @returns {string} Topic: room/{room_id}/agent/{agent_id}/heartbeat
   */
  static heartbeat(roomId, agentId) {
    return `room/${roomId}/agent/${agentId}/heartbeat`
  }

  /**
   * 全局状态 Topic
   * @returns {string} Topic: home/state
   */
  static globalState() {
    return 'home/state'
  }

  /**
   * 策略更新 Topic
   * @returns {string} Topic: home/policy
   */
  static policy() {
    return 'home/policy'
  }

  /**
   * 仲裁 Topic
   * @param {string} [requestId] - 请求 ID（可选，用于响应）
   * @returns {string} Topic
   */
  static arbitration(requestId = null) {
    if (requestId) {
      return `home/arbitration/response/${requestId}`
    }
    return 'home/arbitration'
  }

  /**
   * 系统事件 Topic
   * @returns {string} Topic: home/events
   */
  static events() {
    return 'home/events'
  }
}

/**
 * Topic 解析器
 */
export class TopicParser {
  /**
   * 解析 Topic
   * @param {string} topic - MQTT Topic
   * @returns {Object} 解析结果
   */
  static parse(topic) {
    const parts = topic.split('/')
    const result = {}

    if (parts[0] === 'room' && parts.length >= 2) {
      result.type = 'room'
      result.roomId = parts[1]

      if (parts.length >= 4 && parts[2] === 'agent') {
        result.agentId = parts[3]
      }

      if (parts.length >= 5) {
        result.messageType = parts[4]
      }
    } else if (parts[0] === 'home') {
      result.type = 'home'

      if (parts.length >= 2) {
        result.messageType = parts[1]
      }

      if (parts.length >= 3 && parts[1] === 'arbitration' && parts[2] === 'response') {
        result.isResponse = true
        if (parts.length >= 4) {
          result.requestId = parts[3]
        }
      }
    }

    return result
  }
}

/**
 * QoS 配置
 */
export class QoSConfig {
  static QOS_MAP = {
    'control': 1,
    'state': 0,
    'describe': 1,
    'description': 1,
    'heartbeat': 0,
    'home/state': 0,
    'home/policy': 1,
    'home/arbitration': 1,
    'home/arbitration/response': 1,
    'home/events': 1,
  }

  /**
   * 获取消息类型的 QoS 等级
   * @param {string} messageType - 消息类型
   * @returns {number} QoS 等级
   */
  static getQos(messageType) {
    return this.QOS_MAP[messageType] || 0
  }

  /**
   * 根据 Topic 获取 QoS 等级
   * @param {string} topic - MQTT Topic
   * @returns {number} QoS 等级
   */
  static getQosForTopic(topic) {
    const parsed = TopicParser.parse(topic)

    if (parsed.type === 'home') {
      const messageType = `home/${parsed.messageType || ''}`
      return this.QOS_MAP[messageType] || 0
    }

    return this.QOS_MAP[parsed.messageType] || 0
  }
}

/**
 * 订阅 Topic 模式
 */
export class SubscriptionTopics {
  /**
   * Personal Agent 订阅的 Topics
   * @param {string} roomId - 房间 ID
   * @returns {string[]} Topic 列表
   */
  static personalAgent(roomId) {
    return [
      `room/${roomId}/agent/+/state`,
      `room/${roomId}/agent/+/description`,
      'home/state',
      'home/policy',
      'home/arbitration/response/+',
    ]
  }

  /**
   * Room Agent 订阅的 Topics
   * @param {string} roomId - 房间 ID
   * @param {string} agentId - Agent ID
   * @returns {string[]} Topic 列表
   */
  static roomAgent(roomId, agentId) {
    return [
      `room/${roomId}/agent/${agentId}/control`,
      `room/${roomId}/agent/${agentId}/describe`,
      'home/policy',
    ]
  }

  /**
   * Central Agent 订阅的 Topics
   * @returns {string[]} Topic 列表
   */
  static centralAgent() {
    return [
      'room/+/agent/+/state',
      'room/+/agent/+/heartbeat',
      'home/arbitration',
    ]
  }
}
